common/phal_F4/rcc/rcc.c

```diff
@@ -0,0 +1,320 @@
+
+    // Turn off and wait for PLL to disable
+    RCC->CR &= ~RCC_CR_PLLON;
+    while((RCC->CR & RCC_CR_PLLRDY))
+        ;                                                                                      // Wait for PLL to turn off
+    RCC->PLLCFGR &= ~(RCC_PLLCFGR_PLLSRC_Msk | RCC_PLLCFGR_PLLN_Msk | RCC_PLLCFGR_PLLM_Msk);   // Clear any PLL source, N, and M configuration
+
+    // Select the PLL source and determine a desired input clock rate
+    uint32_t pll_input_f_hz;
+    switch (pll_source)
+    {
+        // 16 MHz HSI
+        case PLL_SRC_HSI16:                                                                              
+            RCC->PLLCFGR |= RCC_PLLCFGR_PLLSRC_HSI;                                            // Select HSI source
+            RCC->CR |= RCC_CR_HSION;                                                           // Turn on HSI
+            while(!(RCC->CR & RCC_CR_HSIRDY))                                                  // Wait for HSI to enable
+                ;
+            pll_input_f_hz = HSI_CLOCK_RATE_HZ;                                                // Set local for PLL input clock rate
+            break;
+    }
+
+    /* Search for a possible PLL configuration */
+    uint8_t pll_input_divisor = 2;                                                             // PLLM
+    uint8_t pll_output_multiplier = 50;                                                        // PLLN
+    bool valid_rate = false;
+    for (; pll_input_divisor <= 63; pll_input_divisor++)                                       // PLLM must be 2 to 63 (Pg. 227)
+    {
+        // VCO input frequency = PLL input clock frequency / PLLM with 2 <= PLLM <= 63
+        uint32_t pll_vco_in_rate = pll_input_f_hz / pll_input_divisor;                         
+        if(pll_vco_in_rate < 1000000 || pll_vco_in_rate > 2000000)                             // VCO input rate must be 1MHz to 2MHz (Pg. 227 PLLM)
+        {
+            continue;
+        }
+        
+        // VCO output frequency = VCO input * PLLN
+        for (; pll_output_multiplier <= 432; pll_output_multiplier ++)                         // PLLN must be 50 to 432 (Pg. 227)
+        {
+            if ((pll_input_f_hz / pll_input_divisor) * pll_output_multiplier == vco_output_rate_target_hz)
+            {
+                valid_rate = true;
+                break;
+            }
+        }
+        if (valid_rate)
+            break;
+    }
+
+    if (!valid_rate)
+        return false;                                                                          // Unable to find a valid clock rate!
+
+    RCC->PLLCFGR |= ((pll_input_divisor) << RCC_PLLCFGR_PLLM_Pos) & RCC_PLLCFGR_PLLM_Msk;      // Set PLLM
+    RCC->PLLCFGR |= ((pll_output_multiplier) << RCC_PLLCFGR_PLLN_Pos) & RCC_PLLCFGR_PLLN_Msk;  // Set PLLN
+
+    // Update global variable used to reference the PLL
+    PLLClockRateHz = (pll_input_f_hz / pll_input_divisor) * pll_output_multiplier;
```

Comment:
Please check if the PLL input divisor set to the register is the actual divisor - 1. This means setting a register value of 0 will divide by 1. See previous version in how it was done for L series.

Reply:
pll_input_divisor is selected from 2 to 63 (PLLM) on F4. It is also mapped value to value instead of off by one as on L4


![divisor_l4](https://github.com/PurdueElectricRacing/firmware/assets/65915981/159657ab-c178-4a7d-937d-aa06edf7a97b)
![divisor](https://github.com/PurdueElectricRacing/firmware/assets/65915981/1273a05a-df41-4859-bc7a-cdcff3964c4a)